### PR TITLE
Find a series held by any boundary condition, not just four listed ones

### LINKED ISSUE #5902: `extract_field_time_series` misses a series held by a custom boundary condition, so its window never advances
A `FieldTimeSeries` with a partly-in-memory backend keeps only a few frames loaded at a time, and
Oceananigans slides that window forward as the clock advances. It does this by calling
`update_field_time_series!` on every series that `extract_field_time_series` finds in the model.

There are three ways to attach a series to a boundary condition: a domain-wide FTS, a boundary
function, and a callable struct. **`extract_field_time_series` finds it in two of them and misses
it in the third**, while `has_field_time_series` — Oceananigans' own check for whether a type
contains a series — says yes to all three.

## Example use case

NumericalEarth's nesting drives a child's lateral boundaries from a parent `FieldTimeSeries`, through a
condition type that stores the series. It samples at the boundary-face node, so it must know which
boundary it is on — it learns that from `regularize_boundary_condition` and bakes it into its type, so
`getbc` dispatches on the dimension. `extract_field_time_series` does not find the parent series behind
such a condition. So `update_field_time_series!` is never called on it, its window never advances, and
the child is driven from whichever frames happened to be loaded at construction.

## Reproduction

```julia
using Oceananigans
using Oceananigans.BoundaryConditions: ValueBoundaryCondition
using Oceananigans.OutputReaders: extract_field_time_series, has_field_time_series

grid  = RectilinearGrid(size = (4, 4, 4), x = (0, 1), y = (0, 1), z = (0, 1),
                        topology = (Bounded, Bounded, Bounded))
times = collect(0.0:1.0:3.0)
fts   = FieldTimeSeries{Center, Center, Center}(grid, times)

plain(v) = ValueBoundaryCondition(v)
west(bc) = CenterField(grid; boundary_conditions =
    FieldBoundaryConditions(west = bc, east = plain(0.0), south = plain(0.0),
                            north = plain(0.0), bottom = plain(0.0), top = plain(0.0)))

@inline read_series(y, z, t, p) = @inbounds p.series[1, 1, 1, 1]

# A callable struct needs no `getbc` method: the fallback at boundary_condition.jl:175 is
# `getbc(condition, args...) = condition(args...)`, so any callable is already a valid condition.
struct HoldsSeries{S}
    series :: S
end
@inline (h::HoldsSeries)(args...) = @inbounds h.series[1, 1, 1, 1]

# Three ways to attach `fts` to the west boundary. The other five sides are plain numbers.
cases = ["the series itself"                           => plain(fts),
         "a boundary function, series in `parameters`" => ValueBoundaryCondition(read_series; parameters = (; series = fts)),
         "a callable struct holding the series"        => plain(HoldsSeries(fts))]

println(rpad("west boundary condition is…", 44), rpad("has_fts", 10), rpad("via the field", 15), "via the BCs")
for (name, bc) in cases
    c = west(bc)
    println(rpad(name, 44),
            rpad(has_field_time_series(typeof(c.boundary_conditions)), 10),
            rpad(length(extract_field_time_series(c)), 15),
            length(extract_field_time_series(c.boundary_conditions)))
end
```

```
west boundary condition is…                 has_fts   via the field  via the BCs
the series itself                           true      1              1
a boundary function, series in `parameters` true      1              1
a callable struct holding the series        true      0              1
```

`has_fts` is `has_field_time_series` on the boundary conditions. The last two columns are the same
call on the same boundary conditions, reached two ways: through the field, and directly. They agree
except in the third case, where going through the field loses the series.

## Why `extract_field_time_series` misses it

`src/OutputReaders/extract_field_time_series.jl`:

```julia
const FTSBC = Union{CPUFTSBC, GPUFTSBC, DFBC, CFBC}         # :65  — the list of four
const FieldBCsFTS = Union{WFTSBCS, EFTSBCS, …}              # :75
const FieldFTS = Field{…, <:FieldBCsFTS} where {…}          # :76

extract_field_time_series(f::FieldFTS) = extract_field_time_series(f.boundary_conditions)   # :78
```

Whether a field's boundary conditions get opened is decided by `FTSBC`: `FieldTimeSeries`,
`GPUAdaptedFieldTimeSeries`, `DiscreteBoundaryFunction`, `ContinuousBoundaryFunction`. That covers the
first two cases above. A `HoldsSeries` is none of them, so the field is not a `FieldFTS`, the method
above never runs, and Julia falls back to:

```julia
CannotPossiblyContainFTS = (:Number, :AbstractArray, :AbstractGrid, :AbstractField, :Returns, :Nothing)
for T in CannotPossiblyContainFTS
    @eval extract_field_time_series(::$T) = ()                                              # :47-51
end
```

`AbstractField` is on that list, so the third case returns empty at the field and its boundary
conditions are never opened.

Nothing further down is broken. `extract_field_time_series(bc::BoundaryCondition)` unwraps `.condition`
(`:89`), and the generic recursion over struct fields (`:36-41`) finds the series inside a
`HoldsSeries`. The field is the only place the recursion stops.

## Why it matters

A series that is never found is never advanced, and that raises no error. It keeps handing back
whichever frames were loaded when it was built, and once the clock runs past the last of them the time
interpolation asks for a frame that is not there:

- **CPU**: nothing is reported. The read returns whatever sits next to the array in memory and the
  run continues with wrong boundary values. With bounds checking on it becomes, for example,
  `BoundsError: attempt to access 12×134×30×2 OffsetArray(…, 1:2) … at index [25, 61, 1, 3]` — every
  spatial index in range, time index 3 against a window holding 2 frames.
- **GPU**: `CUDA error 700`, an illegal memory access, reported far from its cause.

This only happens with a partly-in-memory backend, which is the case you reach for when the series is
too big to hold: the recording that turned this up is 65 GB, and `InMemory(2)` runs it in ~260 MB.

## Suggested fix: finish what #5740 started

#5740 (merged 2026-07-01) added `has_field_time_series` to replace a type-unstable recursion over
`propertynames` with a `@generated` check. It switched the generic path (`:38`), the `Tuple` and
`NamedTuple` paths (`:54`, `:58`) and the `FieldBoundaryConditions` path (`:81`) to the new check, and
left the `Field` method on `FTSBC`. That list is the last piece of the old design still in use, and it
is the piece that gives the wrong answer.

This is not a proposal to undo #5740 — it replaces the one method it skipped:

```julia
extract_field_time_series(f::Field) = extract_field_time_series(f.boundary_conditions)
```

No guard is needed: `extract_field_time_series(bcs::FieldBoundaryConditions)` (`:80`) already begins
with `has_field_time_series(typeof(bcs))` and returns `()` when false, so #5740's compile-time pruning
still happens, one level deeper. `Base.return_types` on a field with no series, and on one with default
boundary conditions, both give `Tuple{}` — a singleton, so the result is settled during inference and
nothing runs.

This replaces the `FieldFTS` method and leaves `FTSBC`, `FieldBCsFTS` and the seven `*FTSBCS` aliases
with no users.

Adding types to `FTSBC` instead would not fix this. Any package downstream can write its own condition
type that stores a series, so the list can never be complete.

Reproduced on Oceananigans 0.110.13 (Julia 1.12.6). Still present on `main` @ 4c09b08e1:
`FTSBC` (`:65`), `FieldFTS` (`:76`), the `AbstractField` entry in `CannotPossiblyContainFTS` (`:47`)
and `extract_field_time_series(f::FieldFTS)` (`:78`) are all unchanged.

### PR DESCRIPTION
**Pull request by ewquon**
Closes #5902.

Whether a field's boundary conditions are opened was decided by `FTSBC`, a list of four condition
types. A condition that holds a `FieldTimeSeries` without being one is not on it, so the series is
never found and its in-memory window never advances — silently on CPU, as `CUDA error 700` on GPU.

Ask `has_field_time_series` instead, the check #5740 introduced and already uses for the generic,
`Tuple`, `NamedTuple` and `FieldBoundaryConditions` paths. A model that owns no series still pays
nothing: the `FieldBoundaryConditions` method begins with that check, so such a field resolves to
`Tuple{}` at compile time and nothing runs.

`FTSBC`, `FieldBCsFTS`, `FieldFTS` and the seven `*FTSBCS` aliases have no remaining users and have
been removed.

The test runs over `archs` and covers all three ways to attach a series to a boundary condition;
reverting the fix fails it. Locally `test_output_readers.jl` (11605) and `test_boundary_conditions.jl`
(312) pass on CPU, and on a Tesla T4 with `TEST_ARCHITECTURE=GPU` (10739 and 312).

### CODE REVIEW

File: src/OutputReaders/extract_field_time_series.jl [hunk 1]
```diff
@@ -58,24 +58,12 @@ extract_field_time_series(t::Union{Tuple, NamedTuple}) =
...
-
-const FieldBCsFTS = Union{WFTSBCS, EFTSBCS, SFTSBCS, NFTSBCS, BFTSBCS, TFTSBCS, IFTSBCS}
-const FieldFTS = Field{LX, LY, LZ, O, G, I, D, T, <:FieldBCsFTS} where {LX, LY, LZ, O, G, I, D, T}
-
-extract_field_time_series(f::FieldFTS) = extract_field_time_series(f.boundary_conditions)
+# A field's boundary conditions may hold a series inside a condition of any type — a `FieldTimeSeries`
+# used directly, a boundary function's `parameters`, or a condition type defined downstream. Ask
+# `has_field_time_series` rather than enumerating condition types: it is `@generated`, so a field whose
+# boundary conditions cannot contain a series still resolves to `()` at compile time. The guard lives in
+# the `FieldBoundaryConditions` method below, so none is needed here.
```
Comment (simone-silvestri) [inline src/OutputReaders/extract_field_time_series.jl]:
```suggestion
# A field's boundary conditions may hold a series inside a condition of any type
```

File: test/test_output_readers.jl [hunk 2]
```diff
@@ -2,11 +2,20 @@ include("dependencies_for_runtests.jl")
...
 using Oceananigans.Fields: indices, interpolate!
-using Oceananigans.OutputReaders: Cyclical, Clamp, Linear, SplitFilePath, cpu_interpolating_time_indices
+using Oceananigans.OutputReaders: Cyclical, Clamp, Linear, SplitFilePath, cpu_interpolating_time_indices,
+                                  extract_field_time_series, has_field_time_series
 
 using Random
 using NCDatasets
 
+# A boundary condition that stores a `FieldTimeSeries` rather than being one. No `getbc` method is
+# needed: the fallback `getbc(condition, args...) = condition(args...)` accepts any callable.
```
Comment (simone-silvestri) [inline test/test_output_readers.jl]:
```suggestion
# A boundary condition that stores a `FieldTimeSeries` rather than being one.
```

File: test/test_output_readers.jl [hunk 3]
```diff
@@ -906,4 +915,46 @@ end
         test_interpolation_with_in_memory_backends(filepath_sine)
     end
     rm(filepath_sine)
+
+    # A series reachable only through a boundary condition must still be found, so that
+    # `update_field_time_series!` advances its in-memory window. A condition that stores a series
+    # without being one used to be missed, because the search consulted a list of condition types
+    # instead of `has_field_time_series`. The window then stayed at the frames it was built with and
+    # the time interpolation eventually indexed past them.
```
Comment (simone-silvestri) [inline test/test_output_readers.jl]:
```suggestion
    # A series reachable only through a boundary condition must still be found, so that
    # `update_field_time_series!` advances its in-memory window.
```

File: test/test_output_readers.jl [hunk 4]
```diff
@@ -906,4 +915,46 @@ end
...
+            @test has_field_time_series(typeof(c.boundary_conditions))
+            @test length(extract_field_time_series(c)) == 1
+            @test first(extract_field_time_series(c)) === fts
+            # Reaching the series through the field must agree with reaching it directly.
+            @test length(extract_field_time_series(c)) ==
+                  length(extract_field_time_series(c.boundary_conditions))
+        end
+
+        # A field that cannot hold a series still resolves to `()` at compile time, so a model owning
+        # no series pays nothing for the check.
```
Comment (simone-silvestri) [inline test/test_output_readers.jl]:
```suggestion
```

**Review by simone-silvestri — APPROVED** [review]
just a trim on extra verbose comments, rest looks good. Historically the limitations on the BCs checked was there because the `extract_field_time_series` function contained type instability and lead to a lot of memory allocation, but with the new `@generated` this is not required anymore. We have a test in place for memory allocation now, so if that test passes, this PR is good to go.
